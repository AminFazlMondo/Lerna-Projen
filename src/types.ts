import { javascript, typescript } from 'projen';

export interface LernaCommonProjectOptions {
  /**
   * (experimental) Flag to run tasks only for the packages that has changes since last release
   *
   * @default false
   * @experimental
   */
  readonly sinceLastRelease?: boolean;

  /**
   * (experimental) Environment variable name to read the git reference from
   * to be used as the base for the `since` option in lerna commands.
   * If not provided, defaults to running git command to find last release tag.
   * The reference that is stored in the environment variable could be:
   * a commit SHA, branch name, or tag name.
   *
   * @experimental
   */
  readonly sinceGitReferenceEnvironmentVariableName?: string;

  /**
   * Whether or not to use Nx for task scheduling
   * https://lerna.js.org/docs/lerna-and-nx
   *
   * @default false
   */
  readonly useNx?: boolean;

  /**
   * Whether or not to use independent versioning for sub-projects
   * https://lerna.js.org/docs/features/version-and-publish#independent-mode
   *
   * @default false
   */
  readonly independentMode?: boolean;

  /**
   * Whether or not to use workspaces in the package.json file
   * Otherwise, will add packages to lerna.json
   *
   * @default false
   */
  readonly useWorkspaces?: boolean;

  /**
   * Key value pair of task name and TaskCustomization to customize the lerna run step added to the task.
   *
   * @default {} "No task customizations"
   */
  readonly taskCustomizations?: TaskCustomizations;

}

export interface LernaProjectOptions extends javascript.NodeProjectOptions, LernaCommonProjectOptions {
  /**
   * (experimental) Use TypeScript for your projenrc file (`.projenrc.ts`).
   *
   * @default false
   * @experimental
   */
  readonly projenrcTs?: boolean;

  /**
   * Consolidate sub projects doc files
   *
   * @default false
   * @experimental
   */
  readonly docgen?: boolean;

  /**
   * (experimental) Docs directory.
   *
   * @default "docs"
   * @experimental
   */
  readonly docsDirectory?: string;
}

export interface LernaTypescriptProjectOptions extends typescript.TypeScriptProjectOptions, LernaCommonProjectOptions {
  /**
   * Whether the project root package has source code.
   *
   * When set to `false`, removes typescript compilation from the compile step for the root project.
   *
   * @default false
   */
  readonly hasRootSourceCode?: boolean;
}

export type TaskCustomizations = { [taskName: string]: TaskCustomization };

export interface TaskCustomization {
  /**
   * Whether to add a lerna run step to the task
   *
   * @default true
   */
  readonly addLernaStep?: boolean;

  /**
   * List of glob patterns matching the name of packages to include in the lerna run step
   *
   * @default []
   */
  readonly include?: string[];

  /**
   * List of glob patterns matching the name of packages to exclude from the lerna run step
   *
   * @default []
   */
  readonly exclude?: string[];

  /**
   * (experimental) Flag to run tasks only for the packages that has changes since last release
   *
   * @default "Value passed in project options"
   */
  readonly sinceLastRelease?: boolean;
}