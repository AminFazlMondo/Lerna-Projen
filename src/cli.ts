import { Command } from 'commander';
import { copy, readdirSync, existsSync, move, emptyDirSync, readFileSync, writeFileSync } from 'fs-extra';

const program = new Command();

interface FileDetailsType {
  path: string;
  content?: Buffer | undefined;
}

program
  .command('clean-dist <distFolder>')
  .action(async (distFolder) => {
    const existingFiles: string[] = [
      `${distFolder}/changelog.md`,
      `${distFolder}/version.txt`,
      `${distFolder}/releasetag.txt`,
    ];
    const fileDetails = existingFiles.map(readFileIfExists);
    emptyDirSync(distFolder);
    fileDetails.forEach(writeFileIfExists);
  });

program
  .command('copy-dist <subProjectDist> <parentDist>')
  .action(async (subProjectPath, parentDist) => {
    const subProjectDist = `./${subProjectPath}`;
    if (!existsSync(subProjectDist)) {return;}

    const entries = readdirSync(subProjectDist);
    await Promise.all(entries.map(f => copy(`${subProjectDist}/${f}`, `${parentDist}/${f}`, { recursive: true, overwrite: true })));
  });


program
  .command('move-docs <parentDocsDirectory> <subProjectPath> <subProjectDocsDirectory>')
  .action(async (parentDocsDirectory, subProjectPath, subProjectDocsDirectory) => {
    await moveDocs(parentDocsDirectory, subProjectPath, subProjectDocsDirectory, false);
    await moveDocs(parentDocsDirectory, subProjectPath, 'API.md', true);
  });

async function moveDocs(parentDocsDirectory: string, subProjectPath: string, subPath: string, isFile: boolean): Promise<void> {
  const subProjectDocs = `./${subProjectPath}/${subPath}`;
  if (!existsSync(subProjectDocs)) {return;}

  const destinationFolder = `./${parentDocsDirectory}/${subProjectPath}`;
  emptyDirSync(destinationFolder);
  const moveOptions = { overwrite: true };
  if (isFile) {
    await move(subProjectDocs, `${destinationFolder}/${subPath}`, moveOptions);
  } else {
    const entries = readdirSync(subProjectDocs);
    await Promise.all(entries.map(f => move(`${subProjectDocs}/${f}`, `${destinationFolder}/${f}`, moveOptions)));
  }
}

function readFileIfExists(filePath: string): FileDetailsType {
  if (!(existsSync(filePath))) {return { path: filePath };}

  return {
    path: filePath,
    content: readFileSync(filePath),
  };
}

function writeFileIfExists(details: FileDetailsType): void {
  if (!details.content) {return;}

  writeFileSync(details.path, details.content);
}

program.parse(process.argv);