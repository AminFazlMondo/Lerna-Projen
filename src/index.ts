import * as path from 'path';
import { JsonFile, javascript, Project, Tasks, SourceCode, typescript, YamlFile } from 'projen';
import { LernaProjectOptions, LernaTypescriptProjectOptions, TaskCustomization, TaskCustomizations } from './types';

export * from './types';
export * as utils from './utils';

function getDocsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'docsDirectory');
  return result?.[1].replace(/\/$/, '');
}

function getArtifactsDirectory(project: Project) {
  const result = Object.entries(project).find(([key]) => key === 'artifactsDirectory');
  return result?.[1];
}

const jsiiTaskPattern = /jsii-docgen -o (?<output>.+)$/i;

function extractJsiiDocsOutput(tasks: Tasks): string | undefined {
  const jsiiDocStep = tasks.tryFind('docgen')?.steps.find(step => jsiiTaskPattern.test(step.exec ?? ''));
  if (!jsiiDocStep || !jsiiDocStep.exec) {return;}

  const match = jsiiDocStep.exec.match(jsiiTaskPattern);
  return match?.groups?.output;
}

function appendWorkflowBootstrapSteps<T extends LernaProjectOptions | LernaTypescriptProjectOptions>(options: T): T {
  if (!options.sinceLastRelease) {return options;}

  return {
    ...options,
    workflowBootstrapSteps: [
      {
        name: 'Fetch tags',
        run: 'if [ $(git rev-parse --is-shallow-repository) != "false" ] ; then git fetch origin main --tags --unshallow; fi',
        if: '${{ github.workflow == \'build\' }}',
      },
      ...options.workflowBootstrapSteps ?? [],
    ],
  };
}

const lockedTaskNames = ['build', 'upgrade', 'upgrade-projen', 'clobber', 'post-upgrade'];

interface ILernaProject {
  readonly sinceLastRelease: boolean;
  readonly useNx: boolean;
  readonly independentMode: boolean;
  readonly useWorkspaces: boolean;
  readonly docsDirectory: string;
  readonly docgen?: boolean;
  readonly taskCustomizations: TaskCustomizations;
  readonly pnpmVersion?: string;

  customizeTask(taskName: string, customization: TaskCustomization): void;
}

/**
 * @pjid lerna-project
 */
export class LernaProject extends javascript.NodeProject implements ILernaProject {

  private projenrcTs: boolean;

  readonly docsDirectory: string;
  readonly docgen: boolean;
  readonly sinceLastRelease: boolean;
  readonly useNx: boolean;
  readonly independentMode: boolean;
  readonly useWorkspaces: boolean;
  readonly taskCustomizations: TaskCustomizations;
  readonly pnpmVersion?: string;

  private readonly factory: LernaProjectFactory;

  constructor(options: LernaProjectOptions) {
    super(appendWorkflowBootstrapSteps(options));

    if (options.projenrcTs) {this.addDevDeps('ts-node', 'typescript');}

    this.docsDirectory = options.docsDirectory ?? 'docs';
    this.docgen = options.docgen ?? false;
    this.sinceLastRelease = options.sinceLastRelease ?? false;
    this.useNx = options.useNx ?? false;
    this.projenrcTs = options.projenrcTs ?? false;
    this.independentMode = options.independentMode ?? false;
    this.useWorkspaces = options.useWorkspaces ?? false;
    this.taskCustomizations = options.taskCustomizations ?? {};
    this.pnpmVersion = options.pnpmVersion;

    this.factory = new LernaProjectFactory(this);
  }

  /**
   * Adds a customization for the specified task
   *
   * @param taskName Name of the task to customize
   * @param customization TaskCustomization options
   */
  customizeTask(taskName: string, customization: TaskCustomization): void {
    this.taskCustomizations[taskName] = customization;
  }

  preSynthesize() {
    super.preSynthesize();
    const projenCommand = this.projenrcTs ? 'ts-node --skip-project .projenrc.ts' : 'node .projenrc.js';
    const { defaultTask } = this;
    if (!defaultTask) {throw new Error('Could not find default task');}
    defaultTask.reset(projenCommand);

    this.factory.build();
  }
}

/**
 * @pjid lerna-ts-project
 */
export class LernaTypescriptProject extends typescript.TypeScriptProject implements ILernaProject {
  readonly sinceLastRelease: boolean;
  readonly useNx: boolean;
  readonly independentMode: boolean;
  readonly useWorkspaces: boolean;
  readonly taskCustomizations: TaskCustomizations;
  readonly pnpmVersion?: string;

  private readonly factory: LernaProjectFactory;

  constructor(options: LernaTypescriptProjectOptions) {
    super(appendWorkflowBootstrapSteps(options));

    this.sinceLastRelease = options.sinceLastRelease ?? false;
    this.useNx = options.useNx ?? false;
    this.independentMode = options.independentMode ?? false;
    this.useWorkspaces = options.useWorkspaces ?? false;
    this.taskCustomizations = options.taskCustomizations ?? {};
    this.pnpmVersion = options.pnpmVersion;

    if (!(options.hasRootSourceCode ?? false)) {this.tasks.tryFind('compile')?.reset();}

    this.tasks.tryFind('docgen')?.reset();

    this.factory = new LernaProjectFactory(this);
  }

  /**
   * Adds a customization for the specified task
   *
   * @param taskName Name of the task to customize
   * @param customization TaskCustomization options
   */
  customizeTask(taskName: string, customization: TaskCustomization): void {
    this.taskCustomizations[taskName] = customization;
  }

  preSynthesize() {
    super.preSynthesize();
    this.factory.build();
  }
}

class LernaProjectFactory {
  constructor(private readonly project: ILernaProject & javascript.NodeProject) {
    project.addDevDeps('lerna-projen', 'lerna');
  }

  build() {
    this.project.packageTask.reset(`mkdir -p ${this.project.artifactsJavascriptDirectory}`);
    this.project.preCompileTask.exec(`lerna-projen clean-dist ${this.project.artifactsDirectory}`);

    this.setupPackageManager();
    this.appendLernaCommands();
    this.addCrossLinks();
    this.updateSubProjects();
    this.addDocumentsIndex();
  }

  private setupPackageManager() {
    if (this.project.package.packageManager === javascript.NodePackageManager.PNPM && Number(this.project.pnpmVersion) >= 9) {
      this.project.npmrc.addConfig('node-linker', 'hoisted');
    }
  }

  private getSubProjectPath(subProject: Project) {
    return path.relative(this.project.outdir, subProject.outdir);
  }

  private addCrossLinks() {

    const lernaConfig: any = {
      useNx: this.project.useNx,
      version: this.project.independentMode ? 'independent' : '0.0.0',
    };

    const packages = this.project.subprojects.map(subProject => this.getSubProjectPath(subProject));

    if (this.project.useWorkspaces) {
      if (this.project.packageManager === javascript.NodePackageManager.PNPM) {
        new YamlFile(this.project, 'pnpm-workspace.yaml', {
          obj: {
            packages,
          },
        });
      }
      this.project.package.addField('workspaces', packages);
    } else {lernaConfig.packages = packages;}


    new JsonFile(this.project, 'lerna.json', {
      obj: lernaConfig,
    });
  }

  private appendLernaCommands() {
    const upgradeTaskName = 'upgrade';
    const postUpgradeTaskName = 'post-upgrade';
    const postUpgradeTask = this.project.tasks.tryFind(postUpgradeTaskName);
    postUpgradeTask?.prependExec(this.getLernaCommand(upgradeTaskName, { sinceLastRelease: false }));
    postUpgradeTask?.exec('npx projen');

    this.project.tasks.all
      .forEach(task => {
        const customization = this.project.taskCustomizations[task.name];
        const addLernaStep = customization?.addLernaStep ?? true;

        if (lockedTaskNames.includes(task.name) || !addLernaStep) {return;}

        task.exec(this.getLernaCommand(task.name, customization));
      });
  }

  private getLernaCommand(taskName: string, customization?: TaskCustomization) {
    const mainCommand = `lerna run ${taskName} --stream`;
    const useSinceFlag = customization?.sinceLastRelease ?? this.project.sinceLastRelease;

    const includePatterns = customization?.include ?? [];
    const excludePatterns = customization?.exclude ?? [];

    const scopeFlags = includePatterns.map((glob) => ` --scope ${glob}`).join('');
    const ignoreFlags = excludePatterns.map((glob) => ` --ignore ${glob}`).join('');


    const postCommand = useSinceFlag ? ' --since $(git describe --abbrev=0 --tags --match "v*" HEAD^)' : '';
    return `${mainCommand}${postCommand}${scopeFlags}${ignoreFlags}`;
  }

  private updateSubProjects() {
    const bumpTask = this.project.tasks.tryFind('bump');
    const unbumpTask = this.project.tasks.tryFind('unbump');

    this.project.subprojects.forEach((subProject) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject);
      const subProjectPath = this.getSubProjectPath(subProject);
      if (this.project.docgen && subProjectDocsDirectory) {this.project.postCompileTask.exec(`lerna-projen move-docs ${this.project.docsDirectory} ${subProjectPath} ${subProjectDocsDirectory}`);}

      const packageAllTask = subProject.tasks.tryFind('package-all');

      if (packageAllTask) {subProject.packageTask.spawn(packageAllTask);}

      const artifactsDirectory = getArtifactsDirectory(subProject);

      this.project.packageTask.exec(`lerna-projen copy-dist ${subProjectPath}/${artifactsDirectory} ${this.project.artifactsDirectory}`);

      subProject.defaultTask?.reset();

      const bumpEnvs = {
        OUTFILE: 'package.json',
        CHANGELOG: `${artifactsDirectory}/changelog.md`,
        BUMPFILE: `${artifactsDirectory}/version.txt`,
        RELEASETAG: `${artifactsDirectory}/releasetag.txt`,
        RELEASE_TAG_PREFIX: '',
      };

      if (bumpTask && !subProject.tasks.tryFind('bump')) {
        const subBumpTask = subProject.addTask(bumpTask.name, {
          description: bumpTask.description,
          condition: bumpTask.condition,
          env: bumpEnvs,
        });
        subBumpTask.builtin('release/bump-version');
      }

      if (unbumpTask && !subProject.tasks.tryFind('unbump')) {
        const subBumpTask = subProject.addTask(unbumpTask.name, {
          description: unbumpTask.description,
          condition: unbumpTask.condition,
          env: bumpEnvs,
        });
        subBumpTask.builtin('release/reset-version');
      }
    });
  }

  private addDocumentsIndex() {
    if (!this.project.docgen) {return;}

    const subProjectsDocs: Record<string, string> = {};

    const indexMarkdown = new SourceCode(this.project, `${this.project.docsDirectory}/index.md`);
    const readmeMarkdown = new SourceCode(this.project, `${this.project.docsDirectory}/README.md`);

    this.project.subprojects.forEach((subProject) => {
      const subProjectDocsDirectory = getDocsDirectory(subProject);
      if (!subProjectDocsDirectory) {return;}

      const subProjectPath = this.getSubProjectPath(subProject);
      const jsiiDocsOutput = extractJsiiDocsOutput(subProject.tasks);
      if (jsiiDocsOutput) {
        const docsPath = `${subProjectPath}/${jsiiDocsOutput}`;
        indexMarkdown.line(`- ## [${subProject.name}](${docsPath})`);
        subProjectsDocs[subProject.name] = docsPath;
      } else {
        subProjectsDocs[subProject.name] = `${subProjectPath}/index.html`;
      }
      readmeMarkdown.line(`- ## [${subProject.name}](../${subProjectPath}/README.md)`);
    });

    const indexHtml = new SourceCode(this.project, `${this.project.docsDirectory}/index.html`);
    indexHtml.line('<!DOCTYPE html>');
    indexHtml.line('<html>');
    indexHtml.open('<body>');
    indexHtml.open('<ul>');
    Object.entries(subProjectsDocs).forEach(([name, docsPath]) => {
      indexHtml.open('<li>');
      indexHtml.open(`<a href="${docsPath}">`);
      indexHtml.line(name);
      indexHtml.close('</a>');
      indexHtml.close('</li>');
    });
    indexHtml.close('</ul>');
    indexHtml.close('</body>');
    indexHtml.line('</html>');
    indexHtml.line();

    this.project.gitattributes.addAttributes(`/${this.project.docsDirectory}/**`, 'linguist-generated');
  }
}