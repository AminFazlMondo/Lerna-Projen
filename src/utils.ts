import { NodeProject } from 'projen/lib/javascript';

function generateNxTargetEntriesForTaskDependency(options: AddNxTaskDependencyOptions) {
  function getDependentOnAttribute() {
    const { dependsOnTaskName, dependsOnProjects } = options;

    if (dependsOnProjects.length === 0 && !dependsOnTaskName) {return undefined;}

    return [{
      projects: dependsOnProjects.length === 0 ? undefined : dependsOnProjects.map(d => d.package.packageName),
      target: dependsOnTaskName,
    }];
  }
  return [
    options.taskName,
    {
      dependsOn: getDependentOnAttribute(),
      cache: options.cache ?? undefined,
      inputs: options.inputs && options.inputs.length > 0 ? options.inputs : undefined,
      outputs: options.outputs && options.outputs.length > 0 ? options.outputs : undefined,
    },
  ];
}

function generateNxConfigForTasksDependency(options: AddNxTaskDependencyOptions[]) {
  return {
    targets: Object.fromEntries(options.map(generateNxTargetEntriesForTaskDependency)),
  };
}

/**
 * Adds dependency to the project for NX task runner
 * see https://nx.dev/reference/project-configuration#dependson
 *
 * @param project The Project to add dependency to
 * @param taskName The task name that is dependent on another tasks
 * @param dependsOnTaskName The task name that is dependent on in other projects
 * @param dependsOnProjects The packages that source project is dependent on
 */
export function addNxTaskDependency(project: NodeProject, taskName: string, dependsOnTaskName: string, ...dependsOnProjects: NodeProject[]): void {
  project.package.addField('nx', generateNxConfigForTasksDependency([{ taskName, dependsOnTaskName, dependsOnProjects }]));
}

function generateNxConfigForProjectDependency(dependsOn: NodeProject[]) {
  return {
    implicitDependencies: dependsOn.map(d => d.package.packageName),
  };
}

/**
 * Adds dependency to the project for NX task runner
 * see https://nx.dev/reference/project-configuration#implicitdependencies
 *
 * @param project The Project to add dependency to
 * @param dependsOn The packages that source project is dependent on
 */
export function addNxProjectDependency(project: NodeProject, ...dependsOn: NodeProject[]): void {
  project.package.addField('nx', generateNxConfigForProjectDependency(dependsOn));
}

export interface AddNxTaskDependencyOptions {
  /**
   * The task name that is dependent on another tasks
   */
  readonly taskName: string;

  /**
   * The task name that is dependent on in other projects
   */
  readonly dependsOnTaskName?: string;

  /**
   * The packages that source project is dependent on
   */
  readonly dependsOnProjects: NodeProject[];

  /**
   * Cache Task Results
   * @default false
   */
  readonly cache?: boolean;

  /**
   * define what gets included as part of the calculated hash (e.g. files, environment variables, etc.)
   */
  readonly inputs?: string[];

  /**
   * define folders where files might be placed as part of the task execution.
   */
  readonly outputs?: string[];
}

export interface AddNxProjectDependencyOptions {
  /**
   * The packages that source project is dependent on
   */
  readonly dependsOnProjects: NodeProject[];
}

export interface AddNxDependencyOptions {
  /**
   * Task dependency options
   * @deprecated use `tasksDependency`
   */
  readonly taskDependency?: AddNxTaskDependencyOptions;

  /**
   * Task dependency options
   */
  readonly tasksDependency?: AddNxTaskDependencyOptions[];

  /**
   * Project dependency options
   */
  readonly projectDependency?: AddNxProjectDependencyOptions;
}

/**
 * Adds dependency to the project for NX task runner
 * see https://nx.dev/reference/project-configuration
 *
 * @param project The Project to add dependency to
 * @param options Dependency options
 */
export function addNxDependency(project: NodeProject, options: AddNxDependencyOptions): void {
  if (options.taskDependency && options.tasksDependency) {throw new Error('Task Dependency and Tasks dependency option cannot be used in conjunction');}

  const projectDependencyConfig =
    options.projectDependency ? generateNxConfigForProjectDependency(options.projectDependency.dependsOnProjects) : undefined;

  const tasksDependency = options.taskDependency ? [options.taskDependency] : options.tasksDependency;

  const taskDependencyConfig =
    tasksDependency ?
      generateNxConfigForTasksDependency(tasksDependency):
      undefined;

  project.package.addField('nx', {
    ...projectDependencyConfig,
    ...taskDependencyConfig,
  });
}
