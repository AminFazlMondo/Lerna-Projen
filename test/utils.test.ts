import { NodeProject } from 'projen/lib/javascript';
import { synthSnapshot } from 'projen/lib/util/synth';
import { mkdtemp } from './util';
import { addNxTaskDependency, addNxProjectDependency, addNxDependency } from '../src/utils';

describe('utils', () => {
  describe('addNxTaskDependency', () => {
    test('should add dependency in nx', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxTaskDependency(project1, 'test', 'lint', project2, project3);
      addNxTaskDependency(project2, 'compile', 'pre-compile', project1);

      const output1 = synthSnapshot(project1);
      const output2 = synthSnapshot(project2);

      expect(output1['package.json'].nx).toMatchSnapshot();
      expect(output2['package.json'].nx).toMatchSnapshot();
    });
  });

  describe('addNxProjectDependency', () => {
    test('should add dependency in nx', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxProjectDependency(project1, project2, project3);
      addNxProjectDependency(project2, project3);

      const output1 = synthSnapshot(project1);
      const output2 = synthSnapshot(project2);

      expect(output1['package.json'].nx).toMatchSnapshot();
      expect(output2['package.json'].nx).toMatchSnapshot();
    });
  });

  describe('addNxDependency', () => {
    test('should add dependency in nx when only task dependency is specified', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxDependency(project1, {
        tasksDependency: [{
          taskName: 'test',
          dependsOnTaskName: 'lint',
          dependsOnProjects: [project2, project3],
        }],
      });

      const output1 = synthSnapshot(project1);

      expect(output1['package.json'].nx).toMatchSnapshot();
    });

    test('should add dependency in nx when only project dependency is specified', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxDependency(project1, {
        projectDependency: {
          dependsOnProjects: [project2, project3],
        },
      });

      const output1 = synthSnapshot(project1);

      expect(output1['package.json'].nx).toMatchSnapshot();
    });

    test('should add dependency in nx when both task and project dependency is specified', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxDependency(project1, {
        projectDependency: {
          dependsOnProjects: [project2],
        },
        tasksDependency: [
          {
            taskName: 'test',
            dependsOnTaskName: 'lint',
            dependsOnProjects: [project3],
          },
          {
            taskName: 'default',
            dependsOnTaskName: 'build',
            dependsOnProjects: [project3],
          },
        ],
      });

      const output1 = synthSnapshot(project1);

      expect(output1['package.json'].nx).toMatchSnapshot();
    });

    test('should add cache flag when specified', () => {
      const project1 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-1',
        outdir: mkdtemp(),
      });
      const project2 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-2',
        outdir: mkdtemp(),
      });
      const project3 = new NodeProject({
        defaultReleaseBranch: 'main',
        name: 'project-3',
        outdir: mkdtemp(),
      });

      addNxDependency(project1, {
        tasksDependency: [
          {
            taskName: 'test',
            dependsOnTaskName: 'lint',
            dependsOnProjects: [project3],
            cache: true,
          },
          {
            taskName: 'default',
            dependsOnTaskName: 'build',
            dependsOnProjects: [project2],
          },
        ],
      });

      const output1 = synthSnapshot(project1);

      expect(output1['package.json'].nx).toMatchSnapshot();
    });

    describe('deprecated', () => {
      test('should add dependency in nx when only task dependency is specified', () => {
        const project1 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-1',
          outdir: mkdtemp(),
        });
        const project2 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-2',
          outdir: mkdtemp(),
        });
        const project3 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-3',
          outdir: mkdtemp(),
        });

        addNxDependency(project1, {
          taskDependency: {
            taskName: 'test',
            dependsOnTaskName: 'lint',
            dependsOnProjects: [project2, project3],
          },
        });

        const output1 = synthSnapshot(project1);

        expect(output1['package.json'].nx).toMatchSnapshot();
      });

      test('should add dependency in nx when both task and project dependency is specified', () => {
        const project1 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-1',
          outdir: mkdtemp(),
        });
        const project2 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-2',
          outdir: mkdtemp(),
        });
        const project3 = new NodeProject({
          defaultReleaseBranch: 'main',
          name: 'project-3',
          outdir: mkdtemp(),
        });

        addNxDependency(project1, {
          projectDependency: {
            dependsOnProjects: [project2],
          },
          taskDependency: {
            taskName: 'test',
            dependsOnTaskName: 'lint',
            dependsOnProjects: [project3],
          },
        });

        const output1 = synthSnapshot(project1);

        expect(output1['package.json'].nx).toMatchSnapshot();
      });
    });
  });
});