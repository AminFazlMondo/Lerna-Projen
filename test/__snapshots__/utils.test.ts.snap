// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils addNxDependency deprecated should add dependency in nx when both task and project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
  ],
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency deprecated should add dependency in nx when only task dependency is specified 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add cache flag when specified 1`] = `
{
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "cache": true,
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add dependency in nx when both task and project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
  ],
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add dependency in nx when only project dependency is specified 1`] = `
{
  "implicitDependencies": [
    "project-2",
    "project-3",
  ],
}
`;

exports[`utils addNxDependency should add dependency in nx when only task dependency is specified 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add inputs when specified 1`] = `
{
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "target": "lint",
        },
      ],
      "inputs": [
        "!{projectRoot}/**/*.md",
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add inputs when specified 2`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "target": "lint",
        },
      ],
      "inputs": [
        "{projectRoot}/src/**",
        "{projectRoot}/test/**",
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add outputs when specified 1`] = `
{
  "targets": {
    "default": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
          ],
          "target": "build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "target": "lint",
        },
      ],
      "outputs": [
        "{projectRoot}/coverage",
      ],
    },
  },
}
`;

exports[`utils addNxDependency should add outputs when specified 2`] = `
{
  "targets": {
    "build": {
      "dependsOn": [
        {
          "target": "pre-build",
        },
      ],
    },
    "test": {
      "dependsOn": [
        {
          "target": "lint",
        },
      ],
      "outputs": [
        "{projectRoot}/dist",
        "{projectRoot}/dist",
      ],
    },
  },
}
`;

exports[`utils addNxProjectDependency should add dependency in nx 1`] = `
{
  "implicitDependencies": [
    "project-2",
    "project-3",
  ],
}
`;

exports[`utils addNxProjectDependency should add dependency in nx 2`] = `
{
  "implicitDependencies": [
    "project-3",
  ],
}
`;

exports[`utils addNxTaskDependency should add dependency in nx 1`] = `
{
  "targets": {
    "test": {
      "dependsOn": [
        {
          "projects": [
            "project-2",
            "project-3",
          ],
          "target": "lint",
        },
      ],
    },
  },
}
`;

exports[`utils addNxTaskDependency should add dependency in nx 2`] = `
{
  "targets": {
    "compile": {
      "dependsOn": [
        {
          "projects": [
            "project-1",
          ],
          "target": "pre-compile",
        },
      ],
    },
  },
}
`;
